infra/gcp/clusters/modules/k8s-infra-gke-cluster/main.tf

```diff
@@ -55,20 +59,141 @@ resource "google_bigquery_dataset" "usage_metering" {
+  }
+
+  // This restricts deletion of this dataset if there is data in it
+  // unique to test_usage_metering
   delete_contents_on_destroy = true
 }
 
 // Create GKE cluster, but with no node pools. Node pools can be provisioned below
-resource "google_container_cluster" "cluster" {
+// Workaround from https://github.com/hashicorp/terraform/issues/22544#issuecomment-582974372
+// to set lifecycle.prevent_destroy to false if is_prod_cluster
+// keep prod_ and test_ identical except for "unique to " comments
+resource "google_container_cluster" "prod_cluster" {
+  count     = var.is_prod_cluster == "true" ? 1 : 0
+  
   name     = var.cluster_name
   location = var.cluster_location
 
   provider = google-beta
   project  = var.project_name
 
   // GKE clusters are critical objects and should not be destroyed
-  // IMPORTANT: should be false on test clusters
+  // unique to prod_cluster
+  lifecycle {
+    prevent_destroy = true
+  }
+
+  // Network config
+  network = "default"
+
+  // Start with a single node, because we're going to delete the default pool
+  initial_node_count = 1
+
+  // Removes the default node pool, so we can custom create them as separate
+  // objects
+  remove_default_node_pool = true
+
+  // Disable local and certificate auth
+  master_auth {
+    username = ""
+    password = ""
+
+    client_certificate_config {
+      issue_client_certificate = false
+    }
+  }
+
+  // Enable google-groups for RBAC
+  authenticator_groups_config {
+    security_group = "gke-security-groups@kubernetes.io"
+  }
+
+  // Enable workload identity for GCP IAM
+  workload_identity_config {
+    identity_namespace = "${var.project_name}.svc.id.goog"
+  }
+
+  // Enable Stackdriver Kubernetes Monitoring
+  logging_service    = "logging.googleapis.com/kubernetes"
+  monitoring_service = "monitoring.googleapis.com/kubernetes"
+
+  // Set maintenance time
+  maintenance_policy {
+    daily_maintenance_window {
+      start_time = "11:00" // (in UTC), 03:00 PST
+    }
+  }
+
+  // Restrict master to Google IP space; use Cloud Shell to access
+  master_authorized_networks_config {
+  }
+
+  // Enable GKE Usage Metering
+  resource_usage_export_config {
+    enable_network_egress_metering = true
+    bigquery_destination {
+      dataset_id = google_bigquery_dataset.prod_usage_metering[0].dataset_id
+    }
+  }
+
+  // Enable GKE Network Policy
+  network_policy {
+    enabled  = true
+    provider = "CALICO"
+  }
+
+  // Configure cluster addons
+  addons_config {
+    horizontal_pod_autoscaling {
+      disabled = false
+    }
+    http_load_balancing {
+      disabled = false
+    }
+    network_policy_config {
+      disabled = false
+    }
+  }
+
+  // Enable PodSecurityPolicy enforcement
+  pod_security_policy_config {
+    enabled = false // TODO: we should turn this on
+  }
+
+  // Enable VPA
+  vertical_pod_autoscaling {
+    enabled = true
+  }
+}
+resource "google_container_cluster" "test_cluster" {
```

Comment:
I'm not sure to understand this resource. Why not define a `module test-k8s-infra-gke-cluster` ?

Reply:
I felt like copy-pasting resource definitions between modules was more likely to fall out-of-sync than copy-pasting within the same module.  Copy-paste is the only approach I can use for any resource whose lifecycle depends on a flag/environment, since terraform doesn't allow these values to be derived from variables.